paddle/fluid/operators/prune_gate_by_capacity_op.cc

```diff
@@ -0,0 +1,122 @@
+    auto expert_count_dims = ctx->GetInputDim("ExpertCount");
+
+    int64_t n_expert = ctx->Attrs().Get<int64_t>("n_expert");
+    int64_t n_worker = ctx->Attrs().Get<int64_t>("n_worker");
+
+    int64_t gate_idx_num_ele = 1;
+    for (int64_t i = 0; i < gate_idx_dims.size(); i++) {
+      gate_idx_num_ele *= gate_idx_dims[i];
+    }
+    int64_t expert_count_num_ele = 1;
+    for (int64_t i = 0; i < gate_idx_dims.size(); i++) {
+      expert_count_num_ele *= expert_count_dims[i];
+    }
+
+    PADDLE_ENFORCE_EQ(
+        gate_idx_num_ele, n_expert * n_worker,
+        platform::errors::Unavailable(
+            "The number of elements for gate_id is ( %ld ) incorrect "
+            "because the number of gate_idx must equal the "
+            "product of n_worker ( %ld ) and n_expert( %ld ). "
+            "Please input appropriate gate_idx again!",
+            gate_idx_num_ele, n_worker, n_expert));
+
+    PADDLE_ENFORCE_EQ(
+        expert_count_num_ele, n_expert * n_worker,
+        platform::errors::Unavailable(
+            "The number of elements for expert_count is ( %ld ) incorrect "
+            "because the number of expert_count must equal the "
+            "product of n_worker ( %ld )and n_expert( %ld ). "
+            "Please input appropriate expert_count again!",
+            expert_count_num_ele, n_worker, n_expert));
+
+    auto in_dims = ctx->GetInputDim("GateIdx");
+    ctx->SetOutputDim("NewGateIdx", in_dims);
+    ctx->SetOutputDim("ExpertCountOut", in_dims);
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
+      const framework::ExecutionContext& ctx) const override {
+    auto data_type = OperatorWithKernel::IndicateVarDataType(ctx, "GateIdx");
+    return framework::OpKernelType(data_type, ctx.device_context());
+  }
+};
+
+class PruneGateByCapacityOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("GateIdx",
+             "(Tensor), The gate_id sequence corresponding to the input data.");
+    AddInput("ExpertCount",
+             "(Tensor), The quantity value counted on the gate_id sequence of "
+             "the input data.");
+    AddAttr<int64_t>("n_expert", "The number of Experts on each worker")
+        .SetDefault(0);
+    AddAttr<int64_t>("n_worker", "The number of workers on the trainer")
+        .SetDefault(0);
+
+    AddOutput("NewGateIdx",
+              "(Tensor), The gate_id sequence corresponding to the new input "
+              "data after passing through prune.");
+    AddOutput(
+        "ExpertCountOut",
+        "(Tensor), The copy quantity value counted on the gate_id sequence of "
+        "the input data.");
+
+    AddComment(R"DOC(
+prune_gate_by_capacity Operator.
+
+This operator is used to prune gate by capacity(CUDA).
+
+)DOC");
+  }
+};
+
+}  // namespace operators
+}  // namespace paddle
+
+namespace ops = paddle::operators;
+
+REGISTER_OPERATOR(prune_gate_by_capacity, ops::PruneGateByCapacityOp,
```

Comment:
Use REGISTER_OP_WITHOUT_GRADIENT instead.

Reply:
> Use REGISTER_OP_WITHOUT_GRADIENT instead.

ok,I will modified it after next commit.